### ISSUE TITLE
plugins: installation errors requires level 3 ruby mastery

### ISSUE THREAD

**Issue by matti** [kontena/kontena#967]
When a plugin installation fails, it looks like this:

```
$ kontena plugin install azure                                                                                    
Installing plugin azure... fail
ERROR:  Error installing kontena-plugin-azure:
    ERROR: Failed to build gem native extension.

    current directory: /Users/jakolehm/.rvm/gems/ruby-2.3.1@jakolehm/gems/nokogiri-1.6.8/ext/nokogiri
/Users/jakolehm/.rvm/rubies/ruby-2.3.1/bin/ruby -r ./siteconf20160904-49238-1ef1wwa.rb extconf.rb
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
    --with-opt-dir
    --with-opt-include
    --without-opt-include=${opt-dir}/include
    --with-opt-lib
    --without-opt-lib=${opt-dir}/lib
    --with-make-prog
    --without-make-prog
    --srcdir=.
    --curdir
    --ruby=/Users/jakolehm/.rvm/rubies/ruby-2.3.1/bin/$(RUBY_BASE_NAME)
/Users/jakolehm/.rvm/gems/ruby-2.3.1@jakolehm/gems/bundler-1.12.5/lib/bundler/rubygems_integration.rb:322:in `block in replace_gem': pkg-config is not part of the bundle. Add it to Gemfile. (Gem::LoadError)
    from extconf.rb:112:in `<main>'

extconf failed, exit code 1

Gem files will remain installed in /Users/jakolehm/.rvm/gems/ruby-2.3.1@jakolehm/gems/nokogiri-1.6.8 for inspection.
Results logged to /Users/jakolehm/.rvm/gems/ruby-2.3.1@jakolehm/extensions/x86_64-darwin-15/2.3.0/nokogiri-1.6.8/gem_make.out
```

Most people will think that the error is `lock in replace_gem': pkg-config is not part of the bundle. Add it to Gemfile. (Gem::LoadError)` or similar, when one should actually:

`$ gem install nokogiri -v '1.6.8'`

and then find a thread in stackoverflow that the fix is to run `xcode-select --install` to get some deps in order and then somehow try to make nokogiri to build.

How could a) dependency checking be better b) errors to be more verbose?

But since fixing a or b won't make plugins "just work"  shouldn't kontena cli be a docker container? I think that this is the only viable option without rewriting kontena-cli... There is docker-for-mac etc that can mount local volumes etc, it should work. Docker is required anyway for kontena to work.

**Comment by matti** [comment]
Might be that `kontena plugin install` is not compatible with RVM / gemsets...?

Now I solved this by adding `kontena-plugin-azure` to the Gemfile..

**Comment by jakolehm** [comment]
> But since fixing a or b won't make plugins "just work" shouldn't kontena cli be a docker container?

Sure cli works in a docker container... but it has some limitations, for example how do we handle vagrant and other stuff that needs access to host.

**Comment by kke** [comment]
I suppose this was the same when installing the kontena-cli package that still included the azure provisioner.

Maybe there is another azure api wrapper or one could be made inside the plugin to lighten the dependencies.

**Comment by matti** [comment]
@jakolehm Is vagrant needed for any kontena "operator" (a person who uses kontena to deploy stuff, eg. not core contributor?) -- wrapping kontena cli in a docker image would make it "just work" ?

**Comment by matti** [comment]
@kke I'd say it's not possible to predict what dependencies will cause problems on random ruby installations... It also greatly saves development time to be able to use whatever gems etc.

**Comment by jakolehm** [comment]
Vagrant is currently the way people test drive kontena. So it must work without too much trouble (or we need to find another way for easy test drives).

**Comment by matti** [comment]
But if kontena-cli is an alternative way to drive kontena-cli? Where vagrant is needed, ruby kontena-cli still works.

What is the actual problem with kontena-cli container vs vagrant? Couldn't kontena-cli in docker be run in host network...?

**Comment by jakolehm** [comment]
Cli container has multiple problems, not only related to the vagrant (just think how `kontena app build` would work). On the other hand, plugins can already use docker (because they are "just" ruby code)... so we could make azure plugin just a wrapper that commands actual azure sdk stuff inside a container.

We are also experimenting with self contained cli installers (built using chef's omnibus). Brave souls can ask for the installer from our slack channel: https://kontenacommunity.slack.com

**Comment by matti** [comment]
Without much thinking: couldn't we pass docker socket to the container? Is there something else that's required for `kontena app build` ?

**Comment by jakolehm** [comment]
Yes we could pass the docker socket (this is the easy part), but we need also pass the host filesystem so that docker cli inside a container can send sources to the docker daemon.

**Comment by matti** [comment]
yesterdays Kontena Meetup discussion summary: filesystem can be passed easily, like the current working directory, but some corner cases might not work (like when referencing kontena.yml from ../../)

I've successfully mounted /Users from my mac to /Users so pretty much everything should work.

Anyway binary installer with bundled gems is a better way to go (that will solve this issue)

**Comment by aleksijohansson** [comment]
I've also encountered the issue of trying to install plugins to the containerised version of the CLI (https://hub.docker.com/r/kontena/cli/). How can one get access to you https://kontenacommunity.slack.com team?

**Comment by matti** [comment]
@aleksijohansson From the kontena README, there is a link

**Comment by aleksijohansson** [comment]
@matti Thanks! Have you considered using docker instead of vagrant for local kontena "stack" (master, agents etc.) setup?

**Comment by matti** [comment]
@aleksijohansson yep, I'm a long time advicate for that. I authored DEVELOPING.md in kontena that describes how to skip vagrant.

**Comment by aleksijohansson** [comment]
@matti Nice, thanks for the info. In order to match the vagrant functionality (multiple nodes) I think we would need to run an instance of docker inside a container and start the agent there. Have you considered that? I spent some time playing around with it, but didn't get it to work yet https://github.com/aleksijohansson/kontena-master

**Comment by matti** [comment]
I think that Docker in Docker causes more harm than good. What about a model where master/mongo is from docker compose and nodes are added as vagrant machines..?

**Comment by aleksijohansson** [comment]
I see the potential of using docker over vagrant in local development of web applications (narrow scope of course) being the speed of the "native" file system compared to nfs or other type of shares between the host machine and the vagrant machine running the application. This benefit should be possible to achieve with kontena master and one kontena node running on the host machine with docker. The benefit is of course only mostly true if the host machine is a linux machine running docker natively. On macOS there should be benefit with docker's own osxfs compared to nfs shares, but it's still having known issues (https://docs.docker.com/docker-for-mac/troubleshoot/#/known-issues).

To sum things up I think that kontena master and one node in host docker provides the speed benefit on local development and vagrant could provide additional nodes into that which would serve another benefit of running/testing the app distribution even locally.

**Comment by matti** [comment]
Could this be closed now with the omnibus installer?

**Comment by jnummelin** [comment]
Yes, let's close this and open more spefic issues if needed
